File: pkg/neonc/compiler.go
package neonc

import (
	// "neon/pkg/enum"
	"neon/pkg/llvm"
)

func Compile(ast AST) {
	target := llvm.NewTarget()
	defer target.Dispose()

	module := target.CreateModule(ast.Name)
	defer module.Dispose()

	ast.Build(module)

	// fn_type := llvm.NewFunctionType(enum.VOID)
	// fn := module.AddFunction("main", fn_type)
	// block := fn.AppendBlock("entry")
	// fn.SetBuilderToBlockEnd(block)
	// fn.AddRet(enum.VOID)

	module.Verify()

	module.Dump()
	target.ModuleToObjectFile(module, "out.o")
}

File: neon/neonc/compiler.cpp
#include "compiler.h"

#include "util/measure.h"
#include "util/cwd.h"
#include "util/read_file.h"
#include "lexer/lexer.h"
#include "parser/parser.h"
#include <neonc.h>
#include "llvm/target.h"

namespace neonc {
    void build(const char * entry) {
        auto measure = Measure();

        auto cwd = get_cwd();

        auto file_path = cwd + "/" + std::string(entry);
        auto file = read_file(file_path);

        auto lexer = Lexer();
        auto tokens = lexer.Tokenize(file_path, file);
        for (auto tok : tokens) tok.dump();
        std::cout << std::endl;

        auto parser = Parser();
        auto ast = parser.parse_ast(file_path, tokens);

        ast.dump();
        std::cout << std::endl;

        auto target = Target();
        auto module = target.create_module(std::string(entry));

        ast.verify();
        ast.build(module);

        module.dump();
        module.verify();

        target.module_to_object_file(module, file_path);

        measure.finish("FINISHED IN:");
    }
}
